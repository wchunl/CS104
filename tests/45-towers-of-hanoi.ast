ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1)
|   |   TOK_TYPE_ID "" (6.22.1)
|   |   |   TOK_VOID "void" (6.22.1)
|   |   |   TOK_IDENT "fail" (6.22.6)
|   |   TOK_PARAM "(" (6.22.11)
|   |   |   TOK_TYPE_ID "" (6.22.12)
|   |   |   |   TOK_STRING "string" (6.22.12)
|   |   |   |   TOK_IDENT "expr" (6.22.19)
|   |   |   TOK_TYPE_ID "" (6.22.25)
|   |   |   |   TOK_STRING "string" (6.22.25)
|   |   |   |   TOK_IDENT "file" (6.22.32)
|   |   |   TOK_TYPE_ID "" (6.22.38)
|   |   |   |   TOK_INT "int" (6.22.38)
|   |   |   |   TOK_IDENT "line" (6.22.42)
|   TOK_FUNCTION "" (6.24.1)
|   |   TOK_TYPE_ID "" (6.24.1)
|   |   |   TOK_VOID "void" (6.24.1)
|   |   |   TOK_IDENT "putchr" (6.24.6)
|   |   TOK_PARAM "(" (6.24.13)
|   |   |   TOK_TYPE_ID "" (6.24.14)
|   |   |   |   TOK_INT "int" (6.24.14)
|   |   |   |   TOK_IDENT "chr" (6.24.18)
|   TOK_FUNCTION "" (6.25.1)
|   |   TOK_TYPE_ID "" (6.25.1)
|   |   |   TOK_VOID "void" (6.25.1)
|   |   |   TOK_IDENT "putint" (6.25.6)
|   |   TOK_PARAM "(" (6.25.13)
|   |   |   TOK_TYPE_ID "" (6.25.14)
|   |   |   |   TOK_INT "int" (6.25.14)
|   |   |   |   TOK_IDENT "num" (6.25.18)
|   TOK_FUNCTION "" (6.26.1)
|   |   TOK_TYPE_ID "" (6.26.1)
|   |   |   TOK_VOID "void" (6.26.1)
|   |   |   TOK_IDENT "putstr" (6.26.6)
|   |   TOK_PARAM "(" (6.26.13)
|   |   |   TOK_TYPE_ID "" (6.26.14)
|   |   |   |   TOK_STRING "string" (6.26.14)
|   |   |   |   TOK_IDENT "str" (6.26.21)
|   TOK_FUNCTION "" (6.28.1)
|   |   TOK_TYPE_ID "" (6.28.1)
|   |   |   TOK_INT "int" (6.28.1)
|   |   |   TOK_IDENT "getchr" (6.28.5)
|   |   TOK_PARAM "(" (6.28.11)
|   TOK_FUNCTION "" (6.29.1)
|   |   TOK_TYPE_ID "" (6.29.1)
|   |   |   TOK_STRING "string" (6.29.1)
|   |   |   TOK_IDENT "getstr" (6.29.8)
|   |   TOK_PARAM "(" (6.29.14)
|   TOK_FUNCTION "" (6.30.1)
|   |   TOK_TYPE_ID "" (6.30.1)
|   |   |   TOK_STRING "string" (6.30.1)
|   |   |   TOK_IDENT "getln" (6.30.8)
|   |   TOK_PARAM "(" (6.30.13)
|   TOK_FUNCTION "" (7.5.1)
|   |   TOK_TYPE_ID "" (7.5.1)
|   |   |   TOK_VOID "void" (7.5.1)
|   |   |   TOK_IDENT "move" (7.5.6)
|   |   TOK_PARAM "(" (7.5.11)
|   |   |   TOK_TYPE_ID "" (7.5.12)
|   |   |   |   TOK_STRING "string" (7.5.12)
|   |   |   |   TOK_IDENT "src" (7.5.19)
|   |   |   TOK_TYPE_ID "" (7.5.24)
|   |   |   |   TOK_STRING "string" (7.5.24)
|   |   |   |   TOK_IDENT "dst" (7.5.31)
|   |   TOK_BLOCK "{" (7.5.36)
|   |   |   TOK_FUNCTION "" (7.6.11)
|   |   |   |   TOK_CALL "(" (7.6.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.6.4)
|   |   |   |   |   TOK_STRINGCON ""Move a disk from the "" (7.6.12)
|   |   |   |   TOK_CALL "(" (7.7.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.7.4)
|   |   |   |   |   TOK_IDENT "src" (7.7.12)
|   |   |   |   TOK_CALL "(" (7.8.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.8.4)
|   |   |   |   |   TOK_STRINGCON "" to the "" (7.8.12)
|   |   |   |   TOK_CALL "(" (7.9.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.9.4)
|   |   |   |   |   TOK_IDENT "dst" (7.9.12)
|   |   |   |   TOK_CALL "(" (7.10.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.10.4)
|   |   |   |   |   TOK_STRINGCON "".\n"" (7.10.12)
|   TOK_FUNCTION "" (7.13.1)
|   |   TOK_TYPE_ID "" (7.13.1)
|   |   |   TOK_VOID "void" (7.13.1)
|   |   |   TOK_IDENT "towers" (7.13.6)
|   |   TOK_PARAM "(" (7.13.13)
|   |   |   TOK_TYPE_ID "" (7.13.14)
|   |   |   |   TOK_INT "int" (7.13.14)
|   |   |   |   TOK_IDENT "ndisks" (7.13.18)
|   |   |   TOK_TYPE_ID "" (7.13.26)
|   |   |   |   TOK_STRING "string" (7.13.26)
|   |   |   |   TOK_IDENT "src" (7.13.33)
|   |   |   TOK_TYPE_ID "" (7.13.38)
|   |   |   |   TOK_STRING "string" (7.13.38)
|   |   |   |   TOK_IDENT "tmp" (7.13.45)
|   |   |   TOK_TYPE_ID "" (7.13.50)
|   |   |   |   TOK_STRING "string" (7.13.50)
|   |   |   |   TOK_IDENT "dst" (7.13.57)
|   |   TOK_BLOCK "{" (7.13.62)
|   |   |   TOK_FUNCTION "" (7.14.4)
|   |   |   |   TOK_IF "if" (7.14.4)
|   |   |   |   |   '<' "<" (7.14.15)
|   |   |   |   |   |   TOK_IDENT "ndisks" (7.14.8)
|   |   |   |   |   |   TOK_INTCON "1" (7.14.17)
|   |   |   |   |   TOK_RETURNVOID "return" (7.14.20)
|   |   |   |   TOK_CALL "(" (7.15.11)
|   |   |   |   |   TOK_IDENT "towers" (7.15.4)
|   |   |   |   |   '-' "-" (7.15.19)
|   |   |   |   |   |   TOK_IDENT "ndisks" (7.15.12)
|   |   |   |   |   |   TOK_INTCON "1" (7.15.21)
|   |   |   |   |   |   TOK_IDENT "src" (7.15.24)
|   |   |   |   |   |   TOK_IDENT "dst" (7.15.29)
|   |   |   |   |   |   TOK_IDENT "tmp" (7.15.34)
|   |   |   |   TOK_CALL "(" (7.16.9)
|   |   |   |   |   TOK_IDENT "move" (7.16.4)
|   |   |   |   |   TOK_IDENT "src" (7.16.10)
|   |   |   |   |   |   TOK_IDENT "dst" (7.16.15)
|   |   |   |   TOK_CALL "(" (7.17.11)
|   |   |   |   |   TOK_IDENT "towers" (7.17.4)
|   |   |   |   |   '-' "-" (7.17.19)
|   |   |   |   |   |   TOK_IDENT "ndisks" (7.17.12)
|   |   |   |   |   |   TOK_INTCON "1" (7.17.21)
|   |   |   |   |   |   TOK_IDENT "tmp" (7.17.24)
|   |   |   |   |   |   TOK_IDENT "src" (7.17.29)
|   |   |   |   |   |   TOK_IDENT "dst" (7.17.34)
|   TOK_FUNCTION "" (7.20.1)
|   |   TOK_TYPE_ID "" (7.20.1)
|   |   |   TOK_INT "int" (7.20.1)
|   |   |   TOK_IDENT "main" (7.20.5)
|   |   TOK_PARAM "(" (7.20.10)
|   |   |   TOK_TYPE_ID "" (7.20.11)
|   |   |   |   TOK_INT "int" (7.20.11)
|   |   |   |   TOK_IDENT "argc" (7.20.15)
|   |   |   TOK_TYPE_ID "" (7.20.21)
|   |   |   |   TOK_ARRAY "array" (7.20.21)
|   |   |   |   |   TOK_TYPE_ID "string" (7.20.27)
|   |   |   |   TOK_IDENT "argv" (7.20.35)
|   |   TOK_BLOCK "{" (7.20.41)
|   |   |   TOK_FUNCTION "" (7.21.4)
|   |   |   |   TOK_BLOCK "{" (7.21.4)
|   |   |   |   |   TOK_FUNCTION "" (7.21.5)
|   |   |   |   |   |   TOK_IF "if" (7.21.5)
|   |   |   |   |   |   |   TOK_NOT "not" (7.21.9)
|   |   |   |   |   |   |   |   TOK_EQ "==" (7.21.19)
|   |   |   |   |   |   |   |   |   TOK_IDENT "argc" (7.21.14)
|   |   |   |   |   |   |   |   |   TOK_INTCON "2" (7.21.22)
|   |   |   |   |   |   |   TOK_CALL "(" (7.21.31)
|   |   |   |   |   |   |   |   TOK_IDENT "fail" (7.21.26)
|   |   |   |   |   |   |   |   TOK_STRINGCON ""argc == 2"" (7.21.32)
|   |   |   |   |   |   |   |   |   TOK_STRINGCON ""45-towers-of-hanoi.oc"" (7.21.45)
|   |   |   |   |   |   |   |   |   TOK_INTCON "21" (7.21.70)
|   |   |   |   ';' ";" (7.21.75)
|   |   |   |   TOK_VARDECL "=" (7.22.14)
|   |   |   |   |   TOK_INT "int" (7.22.4)
|   |   |   |   |   |   TOK_TYPE_ID "count" (7.22.8)
|   |   |   |   |   '-' "-" (7.22.27)
|   |   |   |   |   |   TOK_INDEX "[" (7.22.23)
|   |   |   |   |   |   |   TOK_INDEX "[" (7.22.20)
|   |   |   |   |   |   |   |   TOK_IDENT "argv" (7.22.16)
|   |   |   |   |   |   |   |   TOK_INTCON "1" (7.22.21)
|   |   |   |   |   |   |   TOK_INTCON "0" (7.22.24)
|   |   |   |   |   |   TOK_CHARCON "'0'" (7.22.29)
|   |   |   |   TOK_BLOCK "{" (7.23.4)
|   |   |   |   |   TOK_FUNCTION "" (7.23.5)
|   |   |   |   |   |   TOK_IF "if" (7.23.5)
|   |   |   |   |   |   |   TOK_NOT "not" (7.23.9)
|   |   |   |   |   |   |   |   '>' ">" (7.23.20)
|   |   |   |   |   |   |   |   |   TOK_IDENT "count" (7.23.14)
|   |   |   |   |   |   |   |   |   TOK_INTCON "0" (7.23.22)
|   |   |   |   |   |   |   TOK_CALL "(" (7.23.31)
|   |   |   |   |   |   |   |   TOK_IDENT "fail" (7.23.26)
|   |   |   |   |   |   |   |   TOK_STRINGCON ""count > 0"" (7.23.32)
|   |   |   |   |   |   |   |   |   TOK_STRINGCON ""45-towers-of-hanoi.oc"" (7.23.45)
|   |   |   |   |   |   |   |   |   TOK_INTCON "23" (7.23.70)
|   |   |   |   ';' ";" (7.23.75)
|   |   |   |   TOK_BLOCK "{" (7.24.4)
|   |   |   |   |   TOK_FUNCTION "" (7.24.5)
|   |   |   |   |   |   TOK_IF "if" (7.24.5)
|   |   |   |   |   |   |   TOK_NOT "not" (7.24.9)
|   |   |   |   |   |   |   |   '<' "<" (7.24.20)
|   |   |   |   |   |   |   |   |   TOK_IDENT "count" (7.24.14)
|   |   |   |   |   |   |   |   |   TOK_INTCON "9" (7.24.22)
|   |   |   |   |   |   |   TOK_CALL "(" (7.24.31)
|   |   |   |   |   |   |   |   TOK_IDENT "fail" (7.24.26)
|   |   |   |   |   |   |   |   TOK_STRINGCON ""count < 9"" (7.24.32)
|   |   |   |   |   |   |   |   |   TOK_STRINGCON ""45-towers-of-hanoi.oc"" (7.24.45)
|   |   |   |   |   |   |   |   |   TOK_INTCON "24" (7.24.70)
|   |   |   |   ';' ";" (7.24.75)
|   |   |   |   TOK_CALL "(" (7.25.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.25.4)
|   |   |   |   |   TOK_INDEX "[" (7.25.16)
|   |   |   |   |   |   TOK_IDENT "argv" (7.25.12)
|   |   |   |   |   |   TOK_INTCON "0" (7.25.17)
|   |   |   |   TOK_CALL "(" (7.26.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.26.4)
|   |   |   |   |   TOK_STRINGCON "": "" (7.26.12)
|   |   |   |   TOK_CALL "(" (7.27.11)
|   |   |   |   |   TOK_IDENT "putint" (7.27.4)
|   |   |   |   |   TOK_IDENT "count" (7.27.12)
|   |   |   |   TOK_CALL "(" (7.28.11)
|   |   |   |   |   TOK_IDENT "putstr" (7.28.4)
|   |   |   |   |   TOK_STRINGCON "" disks\n"" (7.28.12)
|   |   |   |   TOK_CALL "(" (7.29.11)
|   |   |   |   |   TOK_IDENT "towers" (7.29.4)
|   |   |   |   |   TOK_IDENT "count" (7.29.12)
|   |   |   |   |   |   TOK_STRINGCON ""source"" (7.29.19)
|   |   |   |   |   |   TOK_STRINGCON ""temporary"" (7.29.29)
|   |   |   |   |   |   TOK_STRINGCON ""distination"" (7.29.42)
|   |   |   |   TOK_RETURN "return" (7.30.4)
|   |   |   |   |   TOK_INTCON "0" (7.30.11)
