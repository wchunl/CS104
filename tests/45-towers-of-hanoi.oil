fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
move:     .function
          .param ptr src
          .param ptr dst
          putstr("Move a disk from the ")
          putstr(src)
          putstr(" to the ")
          putstr(dst)
          putstr(".\n")
          .end
towers:   .function
          .param int ndisks
          .param ptr src
          .param ptr tmp
          .param ptr dst
.if0:     $t0:i = ndisks < 1
          goto .el0 if not $t0:i
.th0:     
.fi0:     $t1:i = ndisks - 1
          towers($t1:i)
          $t2:i = src
          move($t2:i)
          $t3:i = ndisks - 1
          towers($t3:i)
          .end
main:     .function int
          .param int argc
          .param ptr argv
          .local int count
.if1:     $t4:i = argc == 2
          $t5:i = not $t4:i
          goto .el1 if not $t5:i
.th1:     $t6:i = "argc == 2"
          fail($t6:i)
.fi1:     $t7:i = [[0 * :p]
          $t8:i = $t7:i - '0'
          count = $t8:i
.if2:     $t9:i = count > 0
          $t10:i = not $t9:i
          goto .el2 if not $t10:i
.th2:     $t11:i = "count > 0"
          fail($t11:i)
.fi2:     
.if3:     $t12:i = count < 9
          $t13:i = not $t12:i
          goto .el3 if not $t13:i
.th3:     $t14:i = "count < 9"
          fail($t14:i)
.fi3:     $t15:p = argv[0 * :p]
          putstr($t15:i)
          putstr(": ")
          putint(count)
          putstr(" disks\n")
          $t16:i = count
          towers($t16:i)
          return 0
          return 
          .end
