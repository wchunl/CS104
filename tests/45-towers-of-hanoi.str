string_set[   2]:   12597418120160561823 0x11afba8->"int"
string_set[   6]:   11608465765758206812 0x11b5ce8->""distination""
string_set[   8]:   13497578857207247412 0x11ab268->"string"
                    9881412649056685758 0x11b3f08->""argc == 2""
                    4551451650890805270 0x11b3d88->"2"
string_set[  10]:    9201800968376702500 0x11b0c98->"getchr"
string_set[  12]:   14643508047410943861 0x11b39a8->"argv"
string_set[  14]:   15816506932648033945 0x11ab228->"("
                    5026105222461765741 0x11b2118->"ndisks"
                    1396094175042499165 0x11b29e8->"<"
string_set[  15]:   12868445110721718657 0x11afd78->")"
string_set[  17]:   18335579659859433076 0x11b56b8->"": ""
                    6142509188972423790 0x11ab648->""
string_set[  18]:    2899015600721308611 0x11b1608->"21"
string_set[  19]:    2749921394955183836 0x11b5458->"24"
string_set[  20]:    1583576014053361350 0x11b08e8->"str"
string_set[  21]:    9317282556992397945 0x11b2518->"if"
string_set[  22]:   10159970873491820195 0x11b2a88->"1"
string_set[  23]:     232592375681467293 0x11b23b8->"tmp"
string_set[  25]:   11001320821098761033 0x11b1b08->"" to the ""
string_set[  26]:   11597697714117577063 0x11b44b8->"="
                   12291392672316406109 0x11aff08->"line"
                    9611428142580219387 0x11ab668->"fail"
string_set[  27]:   18019194267330817565 0x11b52f8->""count < 9""
string_set[  28]:   18413295556186398326 0x11b0158->"putint"
string_set[  29]:   11911796782241914524 0x11ab248->"void"
string_set[  33]:    6814028903552918164 0x11b37f8->"array"
string_set[  35]:   11868120863286193964 0x11b34c8->"main"
string_set[  41]:   12051031859636400188 0x11b4698->"]"
string_set[  42]:   13686721134187162920 0x11b2068->"}"
string_set[  43]:   14849783028996024063 0x11b36e8->"not"
string_set[  49]:   12679825168387489523 0x11b0768->"num"
                    8559740906699418289 0x11b1a98->""Move a disk from the ""
string_set[  50]:      21606623943142546 0x11b0068->"{"
string_set[  52]:    5344662657899890615 0x11aff28->";"
                   11152315395316048902 0x11b2158->"towers"
string_set[  54]:   11809324432133475174 0x11b1318->"move"
                    7296690267450279874 0x11b03c8->"chr"
string_set[  55]:    2297668033614959926 0x11b46b8->"0"
string_set[  56]:   13258891007726008550 0x11b1808->"dst"
                    3701972582333163920 0x11b3488->"argc"
string_set[  61]:   14168603451105147218 0x11b4e58->"23"
string_set[  63]:   14710271594181063355 0x11b5838->"" disks\n""
                   14526079891305879776 0x11b3938->">"
string_set[  64]:   16579916199511856337 0x11b10f8->"getln"
string_set[  67]:    4706206694200543916 0x11afc68->","
string_set[  71]:   16741722213504744880 0x11b1c88->"".\n""
string_set[  73]:   11100957978370531577 0x11b1578->"src"
                    9386170080689500517 0x11afb38->"expr"
string_set[  76]:   11166701219533395393 0x11b0eb8->"getstr"
string_set[  80]:    7211130878751977441 0x11b0288->"putchr"
string_set[  81]:   11299632471807205110 0x11b5ae8->""source""
string_set[  83]:   13581035385482130682 0x11b45a8->"["
string_set[  84]:    5720712447182791414 0x11b47f8->"'0'"
string_set[  85]:    4017852018407011111 0x11afd58->"file"
                    7372855250431704371 0x11b2d18->"-"
string_set[  87]:    1782657156151550167 0x11b4388->"count"
                   14635384904416252381 0x11b2b28->"return"
string_set[  88]:   15577135225310562476 0x11b5d08->""temporary""
                   13969197794603411352 0x11b4038->""45-towers-of-hanoi.oc""
string_set[  89]:    6623666755989985924 0x11b51a8->"9"
                   13348923631108115385 0x11b0548->"putstr"
string_set[  95]:   14736352883882276958 0x11b3ce8->"=="
string_set[  96]:   15225576330967767377 0x11b4cb8->""count > 0""
load_factor = 0.660
bucket_count = 97
max_bucket_size = 3
