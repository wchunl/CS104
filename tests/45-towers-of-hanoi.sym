fail (6.22.1) {0} void function
   expr (6.22.12) {1} string variable param lval 0
   file (6.22.25) {1} string variable param lval 1
   line (6.22.38) {1} int variable param lval 2

putchr (6.24.1) {1} void function
   chr (6.24.14) {1} int variable param lval 0

putint (6.25.1) {2} void function
   num (6.25.14) {1} int variable param lval 0

putstr (6.26.1) {3} void function
   str (6.26.14) {1} string variable param lval 0

getchr (6.28.1) {4} int function

getstr (6.29.1) {5} string function

getln (6.30.1) {6} string function

move (7.5.1) {7} void function
   src (7.5.12) {1} string variable param lval 0
   dst (7.5.24) {1} string variable param lval 1

towers (7.13.1) {8} void function
   ndisks (7.13.14) {1} int variable param lval 0
   src (7.13.26) {1} string variable param lval 1
   tmp (7.13.38) {1} string variable param lval 2
   dst (7.13.50) {1} string variable param lval 3

main (7.20.1) {9} int function
   argc (7.20.11) {1} int variable param lval 0
   argv (7.20.21) {1} array <string> variable param lval 1
   count (7.22.14) {1} int variable local lval 0
