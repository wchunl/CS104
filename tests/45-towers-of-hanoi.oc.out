; 0.  1: # 1 "45-towers-of-hanoi.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "45-towers-of-hanoi.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "45-towers-of-hanoi.oc" 2
; 7.  4: 
; 7.  5: void move (string src, string dst) {
; 7.  6:    putstr ("Move a disk from the ");
; 7.  7:    putstr (src);
; 7.  8:    putstr (" to the ");
; 7.  9:    putstr (dst);
; 7. 10:    putstr (".\n");
; 7. 11: }
; 7. 12: 
; 7. 13: void towers (int ndisks, string src, string tmp, string dst) {
; 7. 14:    if (ndisks < 1) return;
; 7. 15:    towers (ndisks - 1, src, dst, tmp);
; 7. 16:    move (src, dst);
; 7. 17:    towers (ndisks - 1, tmp, src, dst);
; 7. 18: }
; 7. 19: 
; 7. 20: int main (int argc, array<string> argv) {
; 7. 21:    {if (not (argc == 2)) fail ("argc == 2", "45-towers-of-hanoi.oc", 21);};
; 7. 22:    int count = argv[1][0] - '0';
; 7. 23:    {if (not (count > 0)) fail ("count > 0", "45-towers-of-hanoi.oc", 23);};
; 7. 24:    {if (not (count < 9)) fail ("count < 9", "45-towers-of-hanoi.oc", 24);};
; 7. 25:    putstr (argv[0]);
; 7. 26:    putstr (": ");
; 7. 27:    putint (count);
; 7. 28:    putstr (" disks\n");
; 7. 29:    towers (count, "source", "temporary", "distination");
; 7. 30:    return 0;
; 7. 31: }
hello